File: algorithms/leetcode/0665-non-decreasing-array.py
from typing import List


# 题目链接
# https://leetcode-cn.com/problems/non-decreasing-array/
class NonDecreasingArray:

    def non_decreasing_array(self, nums: List[int]) -> bool:
        # 递减数组指的是每个元素都是一个比一个少(并不是递减1的意思).
        # 数组中只要任意一个成员比上一个成员大, 则不是递减数组.
        #
        # 非递减数组定义: nums[i] <= nums[i+1].
        # 理解: 任意两个元素, 只要值规律是 a <= b, 则视为非递归数组(等于不是, 小于也不是).
        count = 0
        for ni in range(1, len(nums)):          # ni == next_index
            ci = ni - 1                         # ci == current_index
            cn, nn = nums[ci], nums[ni]         # cn == current_num ,  nn == next_num

            if cn > nn:                         # current_num > next_num == 递减数组
                count += 1

                if count > 1:                   # 是递减数组
                    return False

                if ci > 0:
                    pn = nums[ci - 1]           # pn == previous_num

                    if nn < pn:                 # nn 即小于 pn 也小于 cn, 这是一个递减数组
                        nums[ni] = cn           # 这里尝试更改 nn 的值; 采用 cn 的值来, 使 nn == cn, 符合非递减数组.

        return True


def test_1():

    nums = [4, 2, 3]
    nda = NonDecreasingArray()
    ss = nda.non_decreasing_array(nums)
    assert ss is True


def test_2():

    nums = [4, 2, 1]
    nda = NonDecreasingArray()
    ss = nda.non_decreasing_array(nums)
    assert ss is False


def main():

    test_1()
    test_2()


if __name__ == '__main__':
    main()

File: algorithms/leetcode/0665-non-decreasing-array.cpp
#include <iostream>
#include <assert.h>
#include <vector>
#include <algorithm>
#include <numeric>


// 题目链接
// https://leetcode-cn.com/problems/non-decreasing-array/
class NonDecreasingArray {
public:

    bool
    non_decreasing_array(std::vector<int>& nums) {
        int count = 0;

        for (int ni = 1; ni < nums.size(); ni++) {
            int ci = ni - 1;
            int cn = nums[ci], nn = nums[ni];

            if (cn > nn) {
                count++;

                if (count > 1)
                    return false;

                if (ci > 0) {
                    int pn = nums[ci - 1];
                    if (nn < pn)
                        nums[ni] = cn;
                }

            }
        }

        return true;
    }
};


void test_1() {
    std::vector<int> nums = { 4, 2, 3 };
    NonDecreasingArray nda;
    bool ss = nda.non_decreasing_array(nums);
    assert(ss == true);
}


void test_2() {
    std::vector<int> nums = { 4, 2, 1 };
    NonDecreasingArray nda;
    bool ss = nda.non_decreasing_array(nums);
    assert(ss == false);
}


int main(void) {

    test_1();
    test_2();

    return 0;
}

